### PR TITLE
improve(insured-bridge-relayer): Remove deploymentTimestamp hard-coded mapping in favor of web3.eth.getCode check

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

@mrice32 made a great point that there's a much easier way to check if a deposit.quoteTime is before a contract's deployment time without having to store a hardcoded mapping of contract deployment timestamps. The solution is to check the result of `web3.eth.getCode(bridgePoolAddress, blockForDepositQuoteTime)` and if it returns `0x` then the contract did not exist and we can act accordingly.

This PR therefore reverts #3516 in favor of simplifying the relayer logic to use `web3.eth.getCode` instead of the hardcoded `deploymentTimestamp`.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/test/insured-bridge/BridgePool.js [hunk 1]
```diff
@@ -55,6 +55,7 @@ const lpFeeRatePerSecond = toWei("0.0000015");
 const defaultProposerBondPct = toWei("0.05");
 const defaultSlowRelayFeePct = toWei("0.01");
 const defaultInstantRelayFeePct = toWei("0.01");
+const defaultQuoteTimestamp = "100000"; // no validation of this happens on L1.
```
Comment (nicholaspai) [inline packages/core/test/insured-bridge/BridgePool.js:58]: Changes in this file revert #3516

File: packages/core/contracts/insured-bridge/BridgePool.sol [hunk 2]
```diff
@@ -38,11 +38,6 @@ contract BridgePool is Testable, BridgePoolInterface, ERC20, Lockable {
     using FixedPoint for FixedPoint.Unsigned;
     using Address for address;
 
-    // Useful for clients to know if any quote times are for a time prior to the contract's deployment. These
-    // deposits will be impossible to compute realized LP fee % for since there will be no utilization at that quote
-    // time. Therefore, clients need to deal with this special case.
-    uint32 public deploymentTimestamp;
```
Comment (nicholaspai) [inline packages/core/contracts/insured-bridge/BridgePool.sol:44]: Changes in this file revert #3516

File: packages/financial-templates-lib/test/clients/InsuredBridgeL1Client.js [hunk 3]
```diff
@@ -277,8 +277,7 @@ describe("InsuredBridgeL1Client", function () {
       amount: relayAmount,
       slowRelayFeePct: defaultSlowRelayFeePct,
       instantRelayFeePct: defaultInstantRelayFeePct,
-      // quote time must be <= deployment time of BridgePool for relayDeposit to succeed
-      quoteTimestamp: Number((await bridgePool.methods.deploymentTimestamp().call()).toString()),
+      quoteTimestamp: (await web3.eth.getBlock("latest")).timestamp, // set this to the current block timestamp.
```
Comment (nicholaspai) [inline packages/financial-templates-lib/test/clients/InsuredBridgeL1Client.js:280]: Changes in this file revert #3516

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 4]
```diff
@@ -794,16 +800,15 @@ export class Relayer {
...
+    // TODO: We could improve this search by only looking +/- 10 minutes within the quote time, as this is a constraint
+    // enforced by the BridgeDepositBox. The hard part is that determining the average time per L2 block is not trivial.
     else {
-      const bridgePoolDeploymentTime = (
-        await this.l2BlockFinder.getBlockForTimestamp(this.deployTimestamps[relay.l1Token])
-      ).number;
       let blockSearchConfig = {
-        fromBlock: bridgePoolDeploymentTime,
-        toBlock: bridgePoolDeploymentTime + this.l2LookbackWindow,
+        fromBlock: 0,
```
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/Relayer.ts:810]: This is the one part I'm not happy about. This event querying seems expensive in the worst case. Arbitrum team did state that Infura won't have limited event block search configs soonTM so maybe this code can be deprecated shortly, but I'm not sure what the best way to optimize this search is. We could try to look +/- 10 minutes around the deposit.quoteTime but timing on L2 is a dangerous assumption to make.

File: packages/insured-bridge-relayer/src/RelayerConfig.ts [hunk 5]
```diff
@@ -16,22 +16,6 @@ const supportedChainIds = [
   421611, // arbitrum testnet
 ];
 
-// This struct is a hack right now but prevents an edge case bug where a deposit.quoteTime < bridgePool.deployment time.
```
Comment (nicholaspai) [inline packages/insured-bridge-relayer/src/RelayerConfig.ts:19]: Changes in this file revert #3516

**Review by chrismaree — APPROVED** [review]

File: packages/insured-bridge-relayer/src/Relayer.ts (hunk 4 above)
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/Relayer.ts:808]: ye, and I feel that the call to `getCodeDeployedAtTimestamp` is also pretty slow.

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 6]
```diff
@@ -714,6 +713,11 @@ export class Relayer {
     };
   }
 
+  private async getCodeDeployedAtTimestamp(address: string, timestamp: number): Promise<string> {
```
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/Relayer.ts:716]: I really dont like that this has to call `getBlockForTimestamp`. I feel storing a list of constants in the file is better.

Prehaps you should call this _once_ on the initialization of the pool within the L1Client?

Comment (nicholaspai) [reply packages/insured-bridge-relayer/src/Relayer.ts:716]: Hm yeah it's a trade off of more hard coding and required environment variables vs runtime. 

I think what I'll do is open another PR where I just make the changes to the BridgePool and then leave this open as a potential bot "optimization"

**Pull request closed without merging**